### ISSUE TITLE
Proposal: Reset with pending ChildWorkflows

### ISSUE THREAD

**Issue by longquanzheng** [uber/cadence#3914]
**Is your feature request related to a problem? Please describe.**
Right now it's not allowed to reset workflows when the current run has pending childWorkflows or users want to reset to a point that has pending ChildWorkflows. This is painful to use Cadence with ChildWorkflows

There are lots of complains from Cadence users. E.g: https://uber-cadence.slack.com/archives/CL22WDF70/p1610966432021400

**Proposed Solution**

It's super hard to come up with a simple solution to meet all cases. 

The problem with pending childWorkflows is that after reset, the new parent will probably not get the results of the children of old parent. Because they are two different workflows, probably in two different shards, it's very difficult to make an atomic operation. There is some edge case like at the moment of resetting parent, the children are closing, or already closed(but not reported to old parent because old parent probably closed earlier), or even not really started yet. 

Allow passing different options from user when resetting could be the a good idea.
Option #1: Bypass checking and let it be. So new parent will not get the results if they are waiting. 
Option #2: Always starting new children 
Option #3: Best effort to save: If not started then continue to start children, if not finished then let children report to new parent when finish. But if already finished, then new parent will not get the results.
Option #4: Combination of #2 and #3 --- on top of #3, if children already finished, then starting new children.

Overall, #1 is small effort to implement, #2 is medium, #3 and #4 will be quite a lot of work.

**Additional context**
N/A

**Comment by Groxx** [comment]
How exactly do we treat activities that are started-and-running at the point we reset to?  They have the same lifecycle: schedule, then start, then complete, they can fail, and you can reset to "between" any of those.  I assume they eventually time out, as the response will never be accepted?  Or are they eagerly failed?

Either way, it seems like we can treat children-workflows the same way.  In-progress ones are abandoned (their behavior follows ParentClosePolicy), and in principle their Future/attempt resolves (but may be forgotten).  If they have retries (... do we have child-workflow-retry-policy?  I feel like no), a retry is spawned - if not, it isn't.

Children workflows can fail, it's not like this is a new possibility, though this will likely increase exposure to it.  Races between child-complete and parent-abandoning-complete could be considered the same as races with activities: it's a race, there is no wrong decision.  Like any RPC, you can "succeed" but fail to tell your caller.

---

More generally, from thinking about this in the past, I'm kinda convinced that the core problem is that _we have child workflows at all_.

If these were "regular" workflows, and you could get a Future for other workflows, it would both be far more flexible in general and resolve this kind of ambiguity: if the parent is reset, this is just a `WorkflowResetError(new_run_id)`.  Observing-workflows can do whatever they deem appropriate, quite possibly "query observed-workflow, find out what was forgotten, hook up any missing signal channels again".  Starting workflows can be a pseudo-activity, with exactly the same semantics as activities: success/fail/retries, and we already have the duplicate-ID policy to determine what happens when it succeeds more than once.

You can do all of this _today_ with an Activity for your workflow client calls, except for the "get a Future for another workflow" part.  Currently that requires polling to detect all completion scenarios, which isn't particularly scalable.

**Comment by longquanzheng** [comment]
@Groxx Excited that you have those Qs/comments, because you are getting closed to the truth(core of the problem).
My proposal is indeed too brief so hopefully replying to your comments will make them more detailed.

> How exactly do we treat activities that are started-and-running at the point we reset to?

We failed and reschedule new ones. 

> They have the same lifecycle: schedule, then start, then complete
 
True for lifecycle, `Activity: scheduled->started->completed` vs `ChildWF: initiated->started->completed`(The happy cases).
But they are different in details in the server-client protocol, if you look at history events.
he ChildWFStarted event carries the ChildWF runID. In cadence client experience, it's for providing RunID and telling parent workflow that the childWF has started. The code looks something like this:

```go
childFuture, err := Workflow.ExecuteChildWorkflow(...)
runID := childFuture.getRunID() // returns when childWF started
results := childFuture.getResults() // returns when childWF completed
```

> it seems like we can treat children-workflows the same way

That the `Option 2` in my proposal. But the implementation could be more complicated than activity. 
One of the very basic design the reset is that resetting will result a new Run(and new RunID). (We have to do this way to be able to fork workflow history like Git branching). 

So now, if we do reset in the same way as for Activity, the problem is that line of code:

`runID := childFuture.getRunID() // returns when childWF started`

becomes invalid(or potentially). 

So:
* if we can regret and take away that feature, then problem solved. (I kinda  leaning to this if we want to take shortcut, because I don't believe many people are using this runID for business).  --- but this is a breaking change. 
* Or we just annouce that `childFuture.getRunID()` is not guaranteed under reset. That will work. But just that feature becomes less useful. 
* Therefore my proposal is to make a new event type to update the RunID for parent workflow. This is the only solution if we have to keep runID. and this solution is also generic enough for updating RunID in ContinueAsNew case -- which is also broken today.

**Comment by Groxx** [comment]
Retries make "started" ambiguous for activities, as it can happen multiple times, so it's a convenient thing that there is no "activity run ID" / activity-start future exposed.  
But say there was: how would a reset affect that?  I kinda suspect we wouldn't be agonizing over it for so long, and would just introduce a new api for reset-with-retry scenarios :) maybe as a channel of starts instead of a future, or a "GetRunIds()" that always returns the current knowledge of starts. (You could imagine this paired with a "get result for run id X", which is arguably useful as well)


... but child workflows don't have retries, right? So they're just long-running activities like we have now: failing an activity like that seems correct, so why not child workflows? The run id does not need to change, because it can't, and there's no desire to change it: it failed. Start another.

---

I'm half trying to convince myself that I've been over-thinking this for many months (I explicitly avoided using child workflows in my service, because I couldn't find a way to live without reset, and then spent more time trying to figure out how to monitor workflow failures efficiently, which requires child workflows and their futures/contexts...), and half think it's honestly probably a good idea.  We can introduce child-workflow-retry-options to mitigate the pain of failures, same as on activities, and move on.

And in v2, drop the concept in favor of generic external workflows, with simpler semantics. The migration will probably be pretty straightforward for simple uses, and unlocks much more complex uses (M:N observing, for example).

**Comment by longquanzheng** [comment]
> Retries make "started" ambiguous for activities, as it can happen multiple times, so it's a convenient thing that there is no "activity run ID" / activity-start future exposed.

hehe yeah, luckily. Otherwise our reset doesn't work today either. 

> but child workflows don't have retries, right?

They do have retry. (there was a bug that I fixed a few months ago: https://github.com/uber/cadence/issues/3351 ) so you can put a server side backoff retry policy on childWF. 

ChildWF retry will make a new runID too -- meaning that it's also broken already for getting runID from childWfFuture. 

I am not sure if getting results is also broken after retry...I believe so, since getting results from a continuedAsNew childWF has been broken for a long time. @meiliang86  correct me if I am wrong for this one.

**Comment by longquanzheng** [comment]
@Groxx After discussing with you I actually found a simple way to do reset with pending childWFs now. We probably can zoom and talk about it. I may need @meiliang86 to confirm my idea as it require some details about client behavior. 

The proposal is this: on the server, we only provide a slightly modified version of `opt1` of my proposal -- which just allow bypassing the checking of pending childWFs, and also terminating all pending childWFs.

Then in the workflow code, ask the users  to do this:

before resetting a workflow with childWFs, change the code from 
```
childFuture, err := Workflow.ExecuteChildWorkflow(...)
runID := childFuture.getRunID() // returns when childWF started
results := childFuture.getResults() // returns when childWF completed
```
to 
```
childFuture, err := Workflow.ExecuteChildWorkflow(...)
runID := childFuture.getRunID()
isRunIdTerminated := Workflow.ExecuteLocalActivity( checkRunID(runID)
if !isRunIdTerminated {
     // start childWF again because it's terminated by reset
     childFuture, err = Workflow.ExecuteChildWorkflow(...)
      runID = childFuture.getRunID()
}
results := childFuture.getResults() // returns when childWF completed
// and now it's safe to use the runID for business purpose
```

**Comment by longquanzheng** [comment]
> The run id does not need to change, because it can't, and there's no desire to change it: it failed

@Groxx I know why you thought this way. I had the same until I understand the design of Cadence. Workflow history is the key of the design(that's why we called our kernel service "history service"). It uses "event sourcing" data model: https://microservices.io/patterns/data/event-sourcing.html
because of this event sourcing, the very basic design principal/invariance of Cadence: history must be immutable once written. (if following this, you will understand why we have to wait activity complete to write the activity started event, for activity with retry -- a lot of people feel this very hard to understand). 

Therefore, RunID must be changed after reset. Because reset means overriding history by nature. Moreover, we want to keep the history of the previous run, so that we can do reset again -- e.g. for a runIdA of 100 events, the first reset to event 50th, and users realizes this is not good, and they want to reset to event 60th. If we override the event 60th, then we will never be able to reset to 60th.

^ that's why I said our current reset is like "Git" branching. It allows user to fork history.

**Comment by yux0** [comment]
With the new proposal, the checkRunID invokes a remote call so it will be a (system) activity?

If we introduce a version in the start workflow, will it be easier to solve option2?

**Comment by longquanzheng** [comment]
@yux0 
> With the new proposal, the checkRunID invokes a remote call so it will be a (system) activity?

No, it has to be a raw function to invoke RPC, without activity, local activity, or side effect.
If we use any of  activity, local activity, or side effect, the proposal won't work. 
Imagine this: when the parent is waiting for a childWF for the result, and then the parent is reset to a place where the childWF is started. 

If using activity, local activity, or side effect, the parent before resetting will put a marker there, and we will never be able to remove the code. 

However, after reflecting that proposal, I realized  that approach only work for pending childWFs that are not closed. 
But it doesn't work for case that resetting to pending childWFs that are already closed. 

So I come up with another proposal by using StartChildWorkflowExecutionFailedEventAttributes https://github.com/uber/cadence/blob/3a1e1bcd79bb5bbe76c572e41be73fbdb458fe41/proto/public/uber/cadence/api/v1/history.proto#L77 
* When resetting a parent workflow to a point that has pending childWFs, disregarding the current or future of the childWF, always fail the childWFs using StartChildWorkflowExecutionFailedEvent, with additional information that the `failure reason = reset`
* Reset will also terminate all pending workflows in the base run and current run(so that starting new ones won't fail).
* Update client SDK to handle the error when `failure reason = reset` when starting childWF

basically from  
```go
childFuture := Workflow.ExecuteChildWorkflow(...)
runID, err := childFuture.getRunID() // returns when childWF started or started fails
if err != nil{ return err}
results := childFuture.getResults() // returns when childWF completed
```
to 
```go
var childFuture 
for {
    childFuture = Workflow.ExecuteChildWorkflow(...)
    if err = ERR_FAILURE_BY_RESET{
          continue
    }else if err != nil{
          return err
    }
    break
}
runID, err := childFuture.getRunID() // returns when childWF started or started fails
results := childFuture.getResults() // returns when childWF completed
```

**Comment by Groxx** [comment]
> > The run id does not need to change, because it can't, and there's no desire to change it: it failed
> 
> @Groxx I know why you thought this way. I had the same until I understand the design of Cadence. Workflow history is the key of the design(that's why we called our kernel service "history service"). It uses "event sourcing" data model: https://microservices.io/patterns/data/event-sourcing.html
> 
> Because of this event sourcing, the very basic design principal/invariance of Cadence: history must be immutable once written. (if following this, you will understand why we have to wait activity complete to write the activity started event, for activity with retry -- a lot of people feel this very hard to understand. And also we called the internal WF meta data "mutalbeState" because the history is "immutable").

Sure, for a running workflow that's clearly a reasonable choice.  When resetting though we can copy-and-modify, which we already do - the first event now has more (meta)data, pointing to the previous run, and a new run-id.  A reset is _already_ not "just" a fork, even if we describe it as such.

On the server side: do we not copy perhaps?  If it's literally just a pointer to previous + where to read up to, then yeah - this would be a moderately large change on the server.  But it may be one worth making.

> basically from
> 
> ```go
> // normal non-failure-tolerant "start child workflow" code
> ```
> 
> to
> 
> ```go
> // normal failure-tolerant "start child workflow" code
> ```

Yeah, that's basically what I'm thinking.  Any not-yet-completed children just return a new kind of error, workflow code needs to handle it like any other error.  The exposed run-id does make configured-retries annoying though, as ^ doing that effectively means you're losing the code-simplifying niceties of configured-retries.  We could perhaps make this behavior configurable somewhere, e.g. "I do not need the child run-id" -> you can use the simpler form safely.

---

More generally: seems like child-run-id is already flawed, as it'll give you the _first_ run-id, but not _the current one_?  If that's accurate, we could probably consider "child with retries + exposed runid" is a bug, and hide run-id going forward.  The "started" future is still potentially relevant for making sure the request is accepted by the server though.

**Comment by longquanzheng** [comment]
> On the server side: do we not copy perhaps?

Yes, it is seriously hehe. see 
https://github.com/uber/cadence/blob/bd6c19c6930320f37fe2310121a5dc295c9ec2dc/service/history/reset/resetter.go#L323
and
https://github.com/uber/cadence/blob/bd6c19c6930320f37fe2310121a5dc295c9ec2dc/common/persistence/cassandra/cassandraHistoryPersistence.go#L247

Copy is ineffcient for both storage and performance, so we made a big re-architecture on persistence for that(that's why we called it HistoryV2)

> "I do not need the child run-id" -> you can use the simpler form safely

Agreed. I updated my proposal for the same idea.

**Comment by zedongh** [comment]
Any new progress abot this proposal now?

**Comment by fimanishi** [comment]
Will take a look on this in H2 2026
